packages/vite-plugin-tempest/src/plugin.ts

```diff
@@ -0,0 +1,395 @@
+import fs from 'node:fs'
+import type { AddressInfo } from 'node:net'
+import os from 'node:os'
+import { fileURLToPath } from 'node:url'
+import path from 'node:path'
+import colors from 'picocolors'
+import type { Plugin, ResolvedConfig, UserConfig } from 'vite'
+import { loadEnv } from 'vite'
+import type { InputOption } from 'rollup'
+import type { DevelopmentServerUrl, TempestViteConfiguration } from './types'
+import { loadTempestConfiguration } from './config'
+import { isIpv6 } from './utils'
+
+const TEMPEST_ORIGIN_PLACEHOLDER = '__tempest_placeholder__'
+
+let exitHandlersBound = false
+
+export default function tempest(): Plugin {
+	let viteDevServerUrl: DevelopmentServerUrl
+	let bridgeFilePath: string
+	let resolvedConfig: ResolvedConfig
+	let userConfig: UserConfig
+	let tempestConfig: TempestViteConfiguration
+
+	const defaultAliases: Record<string, string> = {
+		'@': '/src',
+	}
+
+	return {
+		name: 'tempest',
+		enforce: 'post',
+		config: async (config, { command, mode }) => {
+			tempestConfig = await loadTempestConfiguration()
+			userConfig = config
+
+			const ssr = !!userConfig.build?.ssr
+			const env = loadEnv(mode, userConfig.envDir || process.cwd(), '')
+			const assetUrl = env.ASSET_URL ?? ''
+			const serverConfig = command === 'serve'
+				? (resolveDevelopmentEnvironmentServerConfig() ?? resolveEnvironmentServerConfig(env))
+				: undefined
+
+			ensureCommandShouldRunInEnvironment(command, env)
+
+			return {
+				base: userConfig.base ?? (command === 'build' ? resolveBase(tempestConfig, assetUrl) : ''),
+				publicDir: userConfig.publicDir ?? false,
+				build: {
+					manifest: userConfig.build?.manifest ?? (ssr ? false : tempestConfig.manifest),
+					outDir: userConfig.build?.outDir ?? resolveOutDir(tempestConfig),
+					rollupOptions: {
+						input: userConfig.build?.rollupOptions?.input ?? resolveInput(tempestConfig),
+					},
+					assetsInlineLimit: userConfig.build?.assetsInlineLimit ?? 0,
+				},
+				server: {
+					origin: userConfig.server?.origin ?? TEMPEST_ORIGIN_PLACEHOLDER,
+					...(serverConfig
+						? {
+								host: userConfig.server?.host ?? serverConfig.host,
+								hmr: userConfig.server?.hmr === false
+									? false
+									: {
+											...serverConfig.hmr,
+											...(userConfig.server?.hmr === true ? {} : userConfig.server?.hmr),
+										},
+								https: userConfig.server?.https ?? serverConfig.https,
+							}
+						: undefined),
+				},
+				resolve: {
+					alias: Array.isArray(userConfig.resolve?.alias)
+						? [
+								...userConfig.resolve?.alias ?? [],
+								...Object.keys(defaultAliases).map((alias) => ({
+									find: alias,
+									replacement: defaultAliases[alias],
+								})),
+							]
+						: {
```

Comment:
Should we use spaces instead of tabs?

Reply:
Should we? I'm in the tabs team, I only use spaces in PHP because of PSR. If you really want to use spaces, we can change the eslint config, but in the JavaScript ecosystem, it's pretty common to use tabs

Reply:
Full disclosure, I'm a spaces guy, but I'm link [this](https://stitcher.io/blog/tabs-are-better) here because it's funny.

Reply:
I was so ready to argue why I think tabs are better... looks like it won't be needed 😼

Reply:
I'm definitely only using spaces to be consistent with PSR :( If it's tabs in JS, then let's keep it as is